File: CreationalPatterns/Prototype/CSharp/Shape.cs
namespace Prototype
{
    public abstract class Shape : IPrototype<Shape>
    {
        public int X { get; set; }
        public int Y { get; set; }
        public string Color { get; set; }

        public Shape() { }
        public Shape(Shape source)
        {
            X = source.X;
            Y = source.Y;
            Color = source.Color;
        }

        public abstract Shape Clone();
    }
}

File: CreationalPatterns/Prototype/TypeScript/Shape.ts
import { Prototype } from './Prototype';

export abstract class Shape implements Prototype<Shape> {
    x: number = 0;
    y: number = 0;
    color: string = '';

    protected constructor(source?: Shape) {
        if (source) {
            this.x = source.x;
            this.y = source.y;
            this.color = source.color;
        }
    }

    abstract clone(): Shape;
}

